Add support for BLAS and LAPACK dependencies

rgommers opened the pull request:
This is a start on implementing the BLAS and LAPACK support discussed in gh-2835.

This PR in its current form passes the added test case, and `dependency('openblas')` works for me now with SciPy in 3 cases beyond `pkg-config`:
- CMake detecting OpenBLAS
- Detecting OpenBLAS installed in the default environment `lib/` and `include/` dirs, but no pkg-config nor CMake installed
- Using a native or cross file to specify the include and library directories for OpenBLAS

This is already nice to have in its current form, however there's a lot more to do. This includes dealing with 32 vs. 64 bit interfaces, supporting CBLAS and (maybe) LAPACKE, and implementing support for other implementations (MKL, Netlib, ATLAS, etc.). The docs I added try to sketch this out - I'm very much not sure about the API to expose here, some feedback would be great.

This is also the first time I'm implementing a new dependency, so I may have taken a few wrong turns. If there's better ways of doing any of this, I'd love to hear them.

eli-schwartz commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,136 @@
+                  methods: T.List[DependencyMethods]) -> T.List['DependencyGenerator']:
+    candidates: T.List['DependencyGenerator'] = []
+
+    if DependencyMethods.CMAKE in methods:
```
You probably want to yield the pkg-config one first, right?

Instead of defining your own `func foo_factory` you can assign `foo_factory = DependencyFactory(name, methods_list, ...)`, which is a simplified API that just creates a factory returning one of each method in the listed order, using the default or overridden name parameters. Example from misc.py:

```python
cups_factory = DependencyFactory(
    'cups',
    [DependencyMethods.PKGCONFIG, DependencyMethods.CONFIG_TOOL, DependencyMethods.EXTRAFRAMEWORK, DependencyMethods.CMAKE],
    configtool_class=CupsDependencyConfigTool,
    cmake_name='Cups',
)
```

eli-schwartz commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,136 @@
+        super().__init__(name, environment, kwargs)
+        self.feature_since = ('0.64.0', '')
+
+        self.env = environment
+        self.name = 'openblas'
+        self.is_found = False
```
All 3 of these are handled in `super()`. Don't worry about it.

eli-schwartz commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,136 @@
+            lib_dir = Path(libdir)
+            if not inc_dir.is_absolute() or not lib_dir.is_absolute():
+                # TODO: why doesn't this fail the build? It gets caught, how to avoid?
+                raise DependencyException('Paths given for openblas_includedir and openblas_librarydir in machine file must be absolute')
```
DependencyException is handled by catching it, converting the error message to:

```python
f'Dependency lookup for {name} with method {c.func.log_tried()!r} failed: {e}'
```

logging it as a debug message, and then re-raising it if and only if the dependency is required.

DependencyException is thus meant to be raised for cases where "the dependency cannot be found", but not necessarily for cases where "your machine file is invalid what did you even do". Some other dependency, perhaps even just a base mesonlib.MesonException, is the best thing to raise if you want it to always fail the build and complain that the configuration is "flat-out wrong".

rgommers replied:
> Some other dependency, perhaps even just a base mesonlib.MesonException, is the best thing to raise if you want it to always fail the build and complain that the configuration is "flat-out wrong".

Thanks, that is what I wanted to achieve indeed. If we can detect that config file content is wrong, I'd much prefer an exception over silently ignoring the problem (it gets logged in `meson-log.txt`, but that's too silent for the average user).

eli-schwartz commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,136 @@
+            raise DependencyException('Both openblas_includedir *and* openblas_librarydir have to be set in your machine file (one is not enough)')
+        else:
+            # We only call this method if incdir or libdir were found
+            raise RuntimeError('Meson internal issue during openblas dependency detection')
```
```
raise mesonlib.MesonBugException('XXXXXXX')
```

yields the error:
```
meson.build:lineno:colno: ERROR: XXXXX

This is a Meson bug and should be reported!
```

It is better than RuntimeError because it explicitly denotes an internal issue.

eli-schwartz commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,136 @@
+            raise RuntimeError('Meson internal issue during openblas dependency detection')
+
+        # Now we have the absolute paths we need - use them:
+        self.link_args += [f'-L{lib_dir}', '-lopenblas', f'-I{inc_dir}']
```
Should we use clib_compiler.find_library here, instead of paired -L and -l flags?

The -I flag should be a self.compile_args.

eli-schwartz commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,136 @@
+        m = re.search(r'\d+(?:\.\d+)+', v)
+        if not m:
+            mlog.debug(f'Failed to extract openblas version information')
+            return '0.0.0'
```
If the version is not found, I would return None, and version checks will simply not work (they cannot) and avoid explicitly logging the version 0.0.0 in the found summary.

self.version is initialized in `super().__init__()` as ` self.version: T.Optional[str] = None`

eli-schwartz reviewed (commented):
Thanks for looking at this! Miscellaneous hodgepodge of review remarks that I found when taking a hurried glance (it's the holiday season right now):

rgommers commented:
Thanks for the quick feedback @eli-schwartz. Enjoy your holiday season!

rgommers commented:
The `UNEXSKIP` is because OpenBLAS isn't installed in CI yet. Before I mess with that, let me ask in which jobs you want me to add it. Maybe the same ones as `hdf5` or `scalapack`, since they're similar scientific computing focused dependencies?

rgommers commented:
As far as I could tell, adding keywords or methods/attributes to a dependency object is not allowed by design, and I should be using `modules` and `.get_variable`. `modules` is not a 100% natural fit, because not everything we're looking for is an optional/add-on component (e.g., the 32/64-bit interface is "give me this one" and symbol suffix is optional data about the library one is building against). However, it seems close enough. 

From that follows the current draft API:
```meson
openblas_dep = dependency('openblas',
  version : '>=0.3.21',
  language: 'c',          # can be c/cpp/fortran
  modules: [
    'interface: ilp64',   # can be lp64 or ilp64 (or auto?)
    'symbol-suffix: 64_', # check/auto-detect? default to 64_ or no suffix?
    'cblas',
  ]
)
# Query properties as needed:
has_cblas = openblas.get_variable('cblas')
is_ilp64 = openblas_dep.get_variable('interface') == 'ilp64'
blas_symbol_suffix = openblas_dep.get_variable('symbol-suffix')
```

It'd be great to get some feedback on this.

I'll probably implement Netlib and MKL as well in this PR, because those are both quite different from OpenBLAS. So if the API works for all three, then that's a good sign.

eli-schwartz commented:
> As far as I could tell, adding keywords or methods/attributes to a dependency object is not allowed by design

I would say returned object methods definitely should be avoided, and really in nearly all cases `.get_variable()` should work or should be *made* to work.

Adding keywords to `dependency()` is more iffy, there are a couple cases where we have pretty boutique kwargs. Here's the full list of kwargs supported by dependency():
https://github.com/mesonbuild/meson/blob/b8e53ed5ea9c2a3eec19396233e646d4c1f667ae/mesonbuild/interpreter/interpreter.py#L252-L272

Many of these are even documented. :rofl: The rest are carefully disclaimed in the docs with:
> This function supports additional [library-specific](https://mesonbuild.com/Dependencies.html#dependencies-with-custom-lookup-functionality) keyword arguments that may also be accepted

I think my personal opinion on the matter is I'd like it if we could avoid new kwargs, but if it would be painful or gross without them, then we might as well add them.

tylerjereddy commented:
Will NumPy need 64-bit ILP64 OpenBLAS support for `3.12` wheels in the next release cycle? I think Sebastian was thinking they might in the meeting today? Not sure if that increases priority here.

rgommers commented:
@tylerjereddy yes, but this one will be quick compared to dealing with the SIMD stuff. I've been too swamped with other topics, but I should revisit this in the near future. We'll probably miss the 1.25.0 release window and will have to do a 1.25.1/2 to complete the switchover, just like for SciPy (IIRC 1.9.3 was the first complete release without `setup.py` parts).

rgommers committed:
Implement BLAS/LAPACK API and dependencies: OpenBLAS, MKL, Accelerate

rgommers committed:
BUG: fix macOS SDK version detection

rgommers committed:
Update docs on BLAS/LAPACK dependencies

[skip ci]

rgommers commented:
Hi all, an update on the status of this: this is pretty much done, and actually already shipped in the Meson that is vendored in NumPy, in the numpy `1.26.1` release. I decided on doing it that way because this has turned out to be quite a complex feature - more so than I had anticipated - and I wanted to make sure to get a good amount of real-world feedback before proposing to merge this. So far we've had one bug report that needs addressing, regarding a Netlib BLAS install where `libblas` and `libcblas` are split libraries.

The docs in this PR contain the API that Meson will expose, most of which was already pre-discussed higher up and in gh-2835. A review of that API now would be quite useful though.

dnicolodi commented:
@rgommers I haven't looked at the implementation, but the the documentation needs some finishing touches: there are several `TBD` notes, some `TODO`s, and some undefined things left.

rgommers commented:
Sorry, I meant "please look at `Dependencies.md` for the API". I have some local changes to fix the one bug reported against NumPy as well as make MyPy happy - and that will address the couple of comments with TODO as well.

rgommers committed:
Improve macOS version check in order to use Accelerate NEWLAPACK

rgommers committed:
Add Netlib BLAS (libblas/libcblas) and LAPACK (liblapack/liblapacke) dependencies

Also a bit of refactoring and improvements to OpenBLAS and MKL logic
to reduce the amount of unnecessary checks.

rgommers committed:
BUG: fix macOS version checks so things work in very old versions

The two things fixed here:
- `xcrun` isn't available on very old versions (MacPorts still supports
  10.5/10.6)
- Version comparison is done correctly now with
  `mesonlib.version_compare`

See https://github.com/numpy/numpy/issues/25406 for the bug report.

drhpc commented:
Can you design this in a way that supports specifying which blas, cblas, lapack, lapacke to use via pkg-config module names? In pkgsrc, we rely on custom pkg-config module names to install differing builds of openblas into the same prefix. You have differing libray names and include directories. The CMake FindBLAS and FindLAPACK (missing FindCBLAS and FindLAPACKE yet, AFAIK) support that via `BLA_PREFER_PKGCONFIG` and `BLA_PKGCONFIG_BLAS `/ `BLA_PKGCONFIG_LAPACK`.

I'd love the same for meson, for all components. Right now we're scratching heads over the scipy build which uses the vendored meson. When using the netlib implementation, which comes as 4 libraries with accompanying pkg-config file, it works for numpy buidl to tell it that `blas=cblas`. It finds libcblas.so, which brings libblas.so in and offers the full API. NumPy uses CBLAS, actually.

But SciPy seems to need BLAS API directly, specifically for its vendored superlu, where a build with `blas=cblas` will result in a `_superlu.so` without any BLAS library linkage (as `-lcblas` is silently dropped because of no used symbols, perhaps?).

I just noticed that `blas=blas` does fix the build, but only because the proposed meson code has a special case for BLAS being named blas and assumes a separate libcblas in that case. this guessing of the library name is dangerous, as you just never know all implementations out there.

Can we have separate choice for CBLAS and LAPACKE pkg-config module names?  The application would have the differing netlib library names and most often all 4  settings being identical (`openblas_openmp`), which might be optimized a bit to clean up duplicated compiler/linker flags, but shouldn't cause trouble in any case. The guesswork to ease the life for users not specifying anything is nice to have for those users, but as packager, I need to be able to be specific and preferrably without diving into the detection logic of each build system as used by each package. I want numerous hours of my life back that I spent hunting down differing FindLAPACK.cmake copies/forks/idiosyncratic variants.

rgommers commented:
> Can you design this in a way that supports specifying which blas, cblas, lapack, lapacke to use via pkg-config module names?

Yes for the main part, as commented on in https://github.com/mesonbuild/meson/issues/2835#issuecomment-1871432452. For separate CBLAS/LAPACKE, see below.

> The CMake FindBLAS and FindLAPACK (missing FindCBLAS and FindLAPACKE yet, AFAIK) support that via `BLA_PREFER_PKGCONFIG` and `BLA_PKGCONFIG_BLAS `/ `BLA_PKGCONFIG_LAPACK`.

It's already the first thing tried in Meson, so nothing else should be needed.

> Right now we're scratching heads over the scipy build which uses the vendored meson.

I'll comment on this elsewhere (your numpy-discussion post probably) since it's SciPy-specific. Just to say here that SciPy, as of now, does _not_ use the code in this PR or a vendored Meson (NumPy does). As a result, it's less flexible. This PR should solve SciPy's issues.

> Can we have separate choice for CBLAS and LAPACKE pkg-config module names?

I'd much prefer not to unless it's absolutely necessary - and I've never come across this need before. There is only one known BLAS implementation AFAIK (namely Netlib) which ships CBLAS symbols as a separate library. And in all cases (including Netlib) a package needs BLAS and CBLAS pointing to the same implementation (`cblas.so` will depend on `blas.so`), so completely separating BLAS from CBLAS doesn't make sense to me.

If you have a case that is not covered here (it sounds like `blas=blas` did), can you please explain in detail what the libraries are and how they are built and named?

drhpc commented:
> It's already the first thing tried in Meson, so nothing else should be needed.

Could you, for the sake of compleness here, mention how meson does implement CMake's `BLA_PKGCONFIG_BLAS` choice if the `-Dblas=myfunkyblas` is a feature of the numpy use of meson, not meson itself (I confused that, as I am rather ignorant on generic meson features, just knowing that there are options defined in projects)? Or do you mean that meson will call CMake's FindBLAS anyway (without arguments that would specify my choice, though)? As I see it, if you want a dependency 'blas', meson will find a pkg-config blas.pc and stop there, despite me wanting openblas_openmp.pc.

Can I tell it, generically, to use myfunkyblas.pc when it encounters a desired dependency named 'blas'? I assumed the`-Dblas=foo` feature was supposed to be something generic, for any named dependency, to choose a certain variant. An ignorant assumption of an nonexisting generalisation?

rgommers commented:
> Could you, for the sake of compleness here, mention how meson does implement CMake's `BLA_PKGCONFIG_BLAS` choice if the `-Dblas=myfunkyblas` is a feature of the numpy use of meson, not meson itself

The description of that is: _"If set, the pkg-config method will look for this module name instead of just `blas`."_ That is not difficult, however it does require something from the package author, because by design Meson avoids use of environment variables, and prefers CLI options. So the package author has to expose a `-Dblas` CLI option. If they do, and the basic implementation in the package's `meson.build` file is to use `dependency(cli_blas_name)`, then using `-Dblas=myfunkyblas` works as expected.

> Can I tell it, generically, to use myfunkyblas.pc when it encounters a desired dependency named 'blas'? I assumed the`-Dblas=foo` feature was supposed to be something generic, for any named dependency, to choose a certain variant. An ignorant assumption of an nonexisting generalisation?

You can't tell it generically to do that indeed. And that's by design, because of: https://mesonbuild.com/Contributing.html#environment-variables

drhpc commented:
Thanks for clarifying. So what I imagined is some kind of default cli by meson along 

 -Dmeson-rename-dep=foo=bar

acting on a dependency named foo in meson.build.

rgommers commented:
> So what I imagined is some kind of default cli by meson along `-Dmeson-rename-dep=foo=bar`

There is no such thing. Looking at this again, the closest things to this are a few dependency-specific env vars that do exist (Boost has 2, CUDA 1: https://mesonbuild.com/Dependencies.html#boost), or CLI flags specific to a module like the `python` or `pkgconfig` one (you can see those if you run `meson configure` on a project).

I guess the thing is that in this PR I'm implementing several new dependencies for BLAS libraries plus a uniform API for treating things like asking for CBLAS or ILP64 - but not a generic "give me any BLAS lib you know about" (yet, at least). I'm not sure if that would be appropriate later on, because there's some things that are a bit opinionated in the NumPy/SciPy way of doing things. It's significantly more flexible than https://cmake.org/cmake/help/latest/module/FindBLAS.html (e.g., setting the search order, looking for a selected set of options, handling MKL threading options, etc.).

There are multiple options here I think:
1. Add "BLAS dependency options" in Meson similarly to "python module options"
2. Add a host of env vars, a la CMake
3. Take what NumPy/SciPy do and put it in a separate git repo that can be easily vendored by other projects, should they want the same full flexibility
4. Leave any CLI flags up to individual packages

Right now we're at (4). I am reasonably sure that (2) is a no-go, given Meson's preference for no env vars and the number of knobs that are possible (NumPy has 8 CLI flags now: https://github.com/numpy/numpy/blob/main/meson_options.txt). (3) or (1) could be done.

For now I'd like to finish this with (4), and then see if it makes sense to go to (3) to share code between NumPy and SciPy, since those are close in behavior.

kloczek commented:
Just free thought .. why at all meson needs such extension if both librariesa are just libraries.
On first look all what is needed here is just add to blas and lapack pkgconfig interface to be able detect them 🤔 
All details of exact library like is it 64 or not typu build is possible to put in pkconfig variables.
Nothing special is in those two libraries which could not be described in .pc files

IMO `scipy` and `numpy` should stick to regular meson + take care of provide .pc files by needed libraries (they are just libraries) instead maintaining own fork of meson. I know that to speedup some computations blas and lapack are used quite often as static libraries however pkgconfig has no any issues to describe static libraries interrace as well (`Requires.provate:`, `Libs.private:` and `Cflags.private:` are dedicated to to describe static library interface).

dnicolodi commented:
It seems that the effort on upstreaming this has stalled. I have the impression that one of the issues is that the `dependecy()` interface is not the best suited to describe all the nuisances of the BLAS and LAPACK implementations desired and the preferred way to look it up. What about moving the look-up to a Meson module? It could work in a way similar to
```meson
blas = import('blas').find_dependency(...)
```
This has the advantage that the `find_dependency()` function can use the most proper arguments and is not limited to what makes sense for a generic dependency. If there is interest in seeing how this approach could work, I'll be happy to work on it.

dnicolodi commented:
Another advantage is that this could easily expose command line options like the ones exposed by the `python` module.

rgommers commented:
> It seems that the effort on upstreaming this has stalled. 

Yeah, that's mostly because I've been quite short on time, after I was happy with the maturity of the code. There are some problems though with integration, primarily that it is impossible to make Mypy happy with the way multiple inheritance is used in the current state of this PR. And I've been dreading a large refactor a bit.

> I have the impression that one of the issues is that the `dependecy()` interface is not the best suited to describe all the nuisances of the BLAS and LAPACK implementations desired and the preferred way to look it up

Yes, the `modules: ` interface is acceptable, but not really optimal. Another annoyance is the caching behavior of `dependency()`, so it's not possible to for example search for `dependency('openblas', modules: ['interface: ilp64'])` and then later try again with say `'interface: lp64'`. The caching keys only on the dependency name, not on the content of `modules` IIRC. That hasn't been a problem in practice for NumPy yet, but it's undesirable behavior and I can see use cases for wanting to try the same dependency name multiple times.

> What about moving the look-up to a Meson module?

Hmm, that's an interesting idea. It would make the API nicer, avoid the caching problem, and probably also make it easier to add an `auto` method that searches multiple BLAS libraries in a given order and with certain criteria. Because there's still quite a bit of code needed to do that nicely now: https://github.com/numpy/numpy/blob/88a6a0070d9d50f034c3f31f5e1d9ceea5f1065e/numpy/meson.build#L71-L155

@eli-schwartz WDYT?

kloczek commented:
> Yeah, that's mostly because I've been quite short on time

But again .. BLAS and LAPACK currently are providing proper pkgconfig and/or cmake module interface
```console
$ rpm -ql blas-devel lapack-devel | egrep -e "cmake|pkgconfig"
/usr/lib64/cmake/cblas-3.12.0
/usr/lib64/cmake/cblas-3.12.0/cblas-config-version.cmake
/usr/lib64/cmake/cblas-3.12.0/cblas-config.cmake
/usr/lib64/cmake/cblas-3.12.0/cblas-targets-relwithdebinfo.cmake
/usr/lib64/cmake/cblas-3.12.0/cblas-targets.cmake
/usr/lib64/cmake/cblas64-3.12.0
/usr/lib64/cmake/cblas64-3.12.0/cblas64-config-version.cmake
/usr/lib64/cmake/cblas64-3.12.0/cblas64-config.cmake
/usr/lib64/cmake/cblas64-3.12.0/cblas64-targets-relwithdebinfo.cmake
/usr/lib64/cmake/cblas64-3.12.0/cblas64-targets.cmake
/usr/lib64/pkgconfig/blas.pc
/usr/lib64/pkgconfig/blas64.pc
/usr/lib64/pkgconfig/cblas.pc
/usr/lib64/pkgconfig/cblas64.pc
/usr/lib64/cmake/lapack-3.12.0
/usr/lib64/cmake/lapack-3.12.0/lapack-config-version.cmake
/usr/lib64/cmake/lapack-3.12.0/lapack-config.cmake
/usr/lib64/cmake/lapack-3.12.0/lapack-targets-relwithdebinfo.cmake
/usr/lib64/cmake/lapack-3.12.0/lapack-targets.cmake
/usr/lib64/cmake/lapack64-3.12.0
/usr/lib64/cmake/lapack64-3.12.0/lapack64-config-version.cmake
/usr/lib64/cmake/lapack64-3.12.0/lapack64-config.cmake
/usr/lib64/cmake/lapack64-3.12.0/lapack64-targets-relwithdebinfo.cmake
/usr/lib64/cmake/lapack64-3.12.0/lapack64-targets.cmake
/usr/lib64/cmake/lapacke-3.12.0
/usr/lib64/cmake/lapacke-3.12.0/lapacke-config-version.cmake
/usr/lib64/cmake/lapacke-3.12.0/lapacke-config.cmake
/usr/lib64/cmake/lapacke-3.12.0/lapacke-targets-relwithdebinfo.cmake
/usr/lib64/cmake/lapacke-3.12.0/lapacke-targets.cmake
/usr/lib64/cmake/lapacke64-3.12.0
/usr/lib64/cmake/lapacke64-3.12.0/lapacke64-config-version.cmake
/usr/lib64/cmake/lapacke64-3.12.0/lapacke64-config.cmake
/usr/lib64/cmake/lapacke64-3.12.0/lapacke64-targets-relwithdebinfo.cmake
/usr/lib64/cmake/lapacke64-3.12.0/lapacke64-targets.cmake
/usr/lib64/pkgconfig/lapack.pc
/usr/lib64/pkgconfig/lapack64.pc
/usr/lib64/pkgconfig/lapacke.pc
/usr/lib64/pkgconfig/lapacke64.pc
/usr/share/man/man3/cmake.3.gz
```
So nothing special on meson side needs to be provided.
Ergo: any type of ABI interface of those libraries can be detected current meson.

BTW it would be good to rename cmake(3) to lapack-cmake(3).

kloczek commented:
After copy grep output just realised that cmake module names are slightly wrong.
Instead `<name>-<version>` it should be just `<name>' because it would allow for example `include(lapack)` in cmake files.

rgommers commented:
@kloczek your input really isn't helpful or to the point. `pkg-config` isn't always available (which is why there are `SystemDependency`'s), nor can it do things like deal with symbol suffixes automatically or selecting something like an ILP64 build in a uniform/sane way across various BLAS libraries. This feature is needed and builds on a lot of real-world experience.

kloczek commented:
> @kloczek your input really isn't helpful or to the point. `pkg-config` isn't always available (which is why there are `SystemDependency`'s), nor can it do things like deal with symbol suffixes automatically or selecting something like an ILP64 build in a uniform/sane way across various BLAS libraries. This feature is needed and builds on a lot of real-world experience.

Additional fact: pkgconfig files and cmake modules are generally about describing API .. not ABI and symbols. SONAME name is lowest level of the description. If you need more use variables in pkgconfig/cmake interfaces ..
If you will accept those facts you may find that ll your concerns bout ABI are a bit off the spot.

ILP64 interface or not is selectable from `<name>`  and `<name>64`. That level of the description is ALREADY present.

eli-schwartz commented:
> Yes, the `modules: ` interface is acceptable, but not really optimal. Another annoyance is the caching behavior of `dependency()`, so it's not possible to for example search for `dependency('openblas', modules: ['interface: ilp64'])` and then later try again with say `'interface: lp64'`. The caching keys only on the dependency name, not on the content of `modules` IIRC. That hasn't been a problem in practice for NumPy yet, but it's undesirable behavior and I can see use cases for wanting to try the same dependency name multiple times.

https://github.com/mesonbuild/meson/blob/b8cdd06af5ecf8ddc9c788c95cf05cb5b8c5cda7/mesonbuild/dependencies/detect.py#L41-L67

eli-schwartz commented:
@kloczek

I offer you fair warning: your input is not going to be considered for this PR, so anything you say is just talking into the void.

drhpc commented:
Regarding symbol suffixes for ILP64 … https://github.com/Reference-LAPACK/lapack/issues/666 got marked as resolved now. I wonder if we are approaching a world where BLAS library symbols are standardized and we could just drop in a library (or two libraries that can be trivially linked together) that offers both 32 bit and 64 bit indices.

As a packager/admin, I just want to stress that whatever magic is implemented in the dependency code of the build system, I need some switch to override the magic and make it just follow a central decision to use _THIS_ _BLAS_ (equivalent to the undesired traditional environment variables like BLAS_LIBS). That's why I added the override to tell CMake to use a specific pkg-config module.

I must admit hat my head is at some distance from the meson/numpy/etc building business right now, but I'll have to work on refreshing a HPC cluster install soon and hope that things are at least not harder than last time. Many users do use pip, conda, or whatever, where this is not my problem, but I'd like to continue to offer central install with a predictable config.

rgommers commented:
> Regarding symbol suffixes for ILP64 … [Reference-LAPACK/lapack#666](https://github.com/Reference-LAPACK/lapack/issues/666) got marked as resolved now. I wonder if we are approaching a world where BLAS library symbols are standardized and we could just drop in a library (or two libraries that can be trivially linked together) that offers both 32 bit and 64 bit indices.

I'd love to see us all get there, but it won't happen any time soon I'm afraid. OpenBLAS will need a lot of work that isn't planned. There's some CMake-only support in Netlib BLAS/LAPACK now I believe, but I don't know if any distros are trying to build like that already.

> As a packager/admin, I just want to stress that whatever magic is implemented in the dependency code of the build system, I need some switch to override the magic and make it just follow a central decision to use THIS BLAS 

Definitely, that's a hard requirement. This PR doesn't even add a 'magic auto-detect' option (yet), and while NumPy and SciPy will have that, the current API for selecting an exact library won't change (e.g., `-Dblas=blas -Dlapack=lapack`, as documented in https://numpy.org/devdocs/building/blas_lapack.html#selecting-specific-blas-and-lapack-libraries).

drhpc commented:
> I'd love to see us all get there, but it won't happen any time soon I'm afraid. OpenBLAS will need a lot of work that isn't planned. There's some CMake-only support in Netlib BLAS/LAPACK now I believe, but I don't know if any distros are trying to build like that already.

Well, pkgsrc is, and perhaps others, as it's just the default on the CMake build. Excerpt from their CMakeLists.txt:
```
option(BUILD_INDEX64 "Build Index-64 API libraries" OFF)
if(BUILD_INDEX64)
  set(BLASLIB "blas64")
  set(CBLASLIB "cblas64")
  set(LAPACKLIB "lapack64")
  set(LAPACKELIB "lapacke64")
  set(TMGLIB "tmglib64")
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -DWeirdNEC -DLAPACK_ILP64 -DHAVE_LAPACK_CONFIG_H")
  set(FORTRAN_ILP TRUE)
else()
  set(BLASLIB "blas")
  set(CBLASLIB "cblas")
  set(LAPACKLIB "lapack")
  set(LAPACKELIB "lapacke")
  set(TMGLIB "tmglib")
endif()
# By default build standard API and extended _64 API
option(BUILD_INDEX64_EXT_API "Build Index-64 API as extended API with _64 suffix" ON)
```

This results in
```
$ nm -D /data/pkg/lib/libblas.so | grep dgemm
0000000000019be0 T dgemm_
000000000006b600 T dgemm_64_
$ nm -D /data/pkg/lib/libblas64.so | grep dgemm
0000000000019730 T dgemm_
00000000000698c0 T dgemm_64_
```
Both 32 bit and 64 bit build (BUILD_INDEX64) contain the added `64_` symbols. (I still patch for library naming, though.)

> Definitely, that's a hard requirement. This PR doesn't even add a 'magic auto-detect' option (yet), and while NumPy and SciPy will have that, the current API for selecting an exact library won't change (e.g., `-Dblas=blas -Dlapack=lapack`, as documented in https://numpy.org/devdocs/building/blas_lapack.html#selecting-specific-blas-and-lapack-libraries).

My only gripe there is the implicit handling of cblas. It works now with our openblas builds including cblas (pkgsrc used to have separate builds of cblas on top of optimized blas) and only netlib having it as separate lib. But it's a funky special case that the 'magic' guesses that it will need to link to libcblas when I tell it that `-Dblas=blas`. Not sure what happens when I try our INDEX64 builds. For those I have that note:
```
# Check if the numpy/scipy ecosystem is ready for full ILP64 before
# allowing/setting BLAS_INDEX64=yes. Scipy needs both variants right now,
# wich does not work without standardized symbol suffix.
```
This might be moot of if we just start using the default build of netlib BLAS with CMake (and possibly with plain Makefile build, too, if someone ports that) now, as indicated above. Major packages making use of the 32 and 64 bit symbols as standardized is a strong case for OpenBLAS and others to add them to their builds. After all, it's not Rocket Science, it's just Computer Science (or Algebra, whatever seems more trivial to you;-)

The needed logic for Meson then just boils down again to just requiring a BLAS and presuming or late checking that it either provides 64 bit symbols (numpy etc. enabling big indices then) or not. Meson ocating a named pkg-config module is just fine, again. Then we can talk about me still building libcblas separately and if that should always be wrapped into one library with libblas, like it is for OpenBLAS, normally.

rgommers commented on mesonbuild/dependencies/blas_lapack.py:
```diff
@@ -0,0 +1,491 @@
+
+    def get_variable(self, **kwargs: T.Dict[str, T.Any]) -> str:
+        # TODO: what's going on with `get_variable`? Need to pick from
+        # cmake/pkgconfig/internal/..., but not system?
```
Note to self: this should be fixed in 1.6.0, see https://mesonbuild.com/Release-notes-for-1-6-0.html#support-for-variable-in-system-dependencies.

freakboy3742 committed:
Add numpy specific extensions for iOS builds.

rgommers commented:
Note to self: the 4 commits I just pushed sync changes from https://github.com/numpy/meson that we've been using in NumPy for over a year now. Things are back in sync with that. In addition, I had an attempt at trying to refactor to make Mypy happy at https://github.com/rgommers/meson/tree/make-mypy-happy (didn't look at this branch now, it may or may not be in the right direction).

rgommers commented:
Superseded by gh-14773, closing this PR.
